design/overview.md

```diff
@@ -0,0 +1,145 @@
+- Supports submitting unit file definitions
+- Supports querying against the state of the unit file definition being processed
+- Supports users viewing the generated unit files
+- Fleet is source of truth - no state is held in ```formicad``` that can not be regenerated from fleet
+- Uses a task system with swappable backends
+- No user management, authentication or authorization
+
+### ```formicactl```
+- Supports sending unit file definitions to ```formicad```
+- Supports querying state of unit file definition being processed
+
+## Example Flow
+
+### Using command line tool
+- User submits unit file definition to service using command line tool
+- Service validates unit file definition, returns 200
+- Service generates unit files from unit file definition
+- Service submits all unit files to fleet cluster
+- Meanwhile, command line tool periodically queries service for unit file definition status, displays status to user
+
+## Shipping
+- Docker image as main deployment artifact
+- Continuous integration with automated unit tests
+- Continuous deployment to build Docker image and push image to Docker Hub
+
+## Unit File Definition
+This shows an example of and describes the unit file definition.
+```
+$ cat example.json
+{
+  "units": {
+    "ambassador.service": {
+      "includes": ["ambassador.json", "restart_policy.json"],
+      "overrides": {
+        "Unit": {
+          "Wants": ["user-app.service]
+        },
+      }
+    },
+    "user-app.service": {
+      "includes": ["restart_policy.json"],
+      "systemd": {
+        "Unit": {
+          "Description": "description",
+          "Wants": [
+            "ambassador.service",
+            "lb-register.service"
+          ],
+          "After": [
+            "ambassador.service"
+          ]
+        },
+        "Service": {
+          "ExecStart": "docker run..."
+        },
+        "X-Fleet": {
+          "Global": true
+        }
+      }
+    },
+    "lb-register.service": {
+      "includes": ["restart_policy.json"],
+      "systemd": {
+        "Unit": {
+          "Description": "description",
+          "Wants": [
+            "user-app.service"
+          ],
+          "After": [
+            "user-app.service"
+          ]
+        },
+        "Service": {
+          "ExecStart": "docker run..."
+        },
+        "X-Fleet": {
+          "Global": true
+        }
+      }
+    }
+  }
+}
+
+$ cat ambassador.json
+{
+  "systemd": {
+    "Unit": {
+      "Description": "description",
+      "Wants": []
+    },
+    "Service": {
+      "ExecStart": "docker run..."
+    },
+    "X-Fleet": {
+      "Global": true
+    }
+  }
+}
+
+$ cat restart_policy.json
+{
+  "Restart": "on-failure",
+  "RestartSec": "1s",
+  "StartLimitInterval": "300s",
+  "StartLimitBurst": "3",
+  "TimeoutStartSec": "0"
+}
+```
+Units are described by the "systemd" key, whose keys map directly to systemd unit values, allowing new systemd unit values without re-releasing
```

Comment:
As in, we don't have a mapping between keys in the unit file definition and systemd unit values in ```formicad```.

```
"X-Fleet": {
   "MachineMetadata="role-core=true"
}
```
is translated directly to
```
[X-Fleet]
MachineMetadata="role-core=true"
```

This means we don't have to add support for "X-Whatever" or "SomeOtherValue" with a new release, they're copied directly into the unit file.